### PR TITLE
HTTPCORE-756  - Enforced the requirement that `https` and `http` URIs must not have an empty host identifier

### PR DESCRIPTION
**Pull request by arturobernalg**
This PR brings our `URIBuilder` in line with the standards set by RFC 9110, specifically Section 4.2.2 regarding HTTPS URIs.

Changes:

- Implemented a check to ensure `https` and `http` URIs do not have an empty host identifier.
[RFC](https://www.rfc-editor.org/rfc/rfc9110#name-https-uri-scheme) mandates such URIs be rejected as invalid.

### CODE REVIEW

**Comment by ok2c** [comment]
@arturobernalg Is this PR ready for review?

**Comment by arturobernalg** [comment]
> @arturobernalg Is this PR ready for review?

@ok2c yes. Or is there something I'm missing?

File: httpcore5/src/main/java/org/apache/hc/core5/net/URIBuilder.java [hunk 1]
```diff
@@ -313,11 +314,14 @@ public URI build() throws URISyntaxException {
         return new URI(buildString());
     }
 
-    private String buildString() {
+    private String buildString() throws URISyntaxException {
         final StringBuilder sb = new StringBuilder();
         if (this.scheme != null) {
             sb.append(this.scheme).append(':');
         }
+        if (URIScheme.HTTPS.same(scheme) && (host == null || host.isEmpty())) {
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/net/URIBuilder.java:322]: @arturobernalg Could you please use `TextUtils#isBlank` here for the host check? Looks good otherwise.

**Review by ok2c — APPROVED** [review]

File: httpcore5/src/main/java/org/apache/hc/core5/net/URIBuilder.java (hunk 1 above)
Comment (arturobernalg) [reply httpcore5/src/main/java/org/apache/hc/core5/net/URIBuilder.java]: @ok2c you're right. Changed.

File: httpcore5/src/test/java/org/apache/hc/core5/concurrent/TestBasicFuture.java [hunk 2]
```diff
@@ -260,41 +260,41 @@ public void cancelled() {
 
     @Test
     void testGetWithTimeout() {
+        final AtomicBoolean isFutureCompleted = new AtomicBoolean(false);
```
Comment (ok2c) [inline httpcore5/src/test/java/org/apache/hc/core5/concurrent/TestBasicFuture.java:263]: @arturobernalg Looks good, but there appears to be unrelated change in the change-set. Please move it into a separate commit or a separate PR.

**Review by ok2c — CHANGES_REQUESTED** [review]

File: httpcore5/src/main/java/org/apache/hc/core5/net/URIBuilder.java [hunk 3]
```diff
@@ -309,11 +310,14 @@ public URI build() throws URISyntaxException {
         return new URI(buildString());
     }
 
-    private String buildString() {
+    private String buildString() throws URISyntaxException {
         final StringBuilder sb = new StringBuilder();
         if (this.scheme != null) {
             sb.append(this.scheme).append(':');
         }
+        if (URIScheme.HTTPS.same(scheme) && (TextUtils.isBlank(host))) {
```
Comment (reschke) [inline httpcore5/src/main/java/org/apache/hc/core5/net/URIBuilder.java:318]: Why is this specific to HTTPS?

Comment (arturobernalg) [reply httpcore5/src/main/java/org/apache/hc/core5/net/URIBuilder.java:318]: True. should be `HTTP` as well.

File: httpcore5/src/test/java/org/apache/hc/core5/concurrent/TestBasicFuture.java (hunk 2 above)
Comment (arturobernalg) [reply httpcore5/src/test/java/org/apache/hc/core5/concurrent/TestBasicFuture.java]: @ok2c. fair enough. Changed.

**Comment by reschke** [comment]
I'm not sure that changing `toString()` is a good idea here.

**Comment by ok2c** [comment]
@arturobernalg What you can do here is move the checks to the `#build` method and keep `#buildString` as is. That would also enable to remove that ugly try-catch from the `#toString`

**Comment by ok2c** [comment]
@arturobernalg Please also do not forget to update the description as it will be used as a commit message

**Comment by arturobernalg** [comment]
> @arturobernalg Please also do not forget to update the description as it will be used as a commit message

@ok2c 
Done

**Review by ok2c — APPROVED** [review]
